integration_tests/tests/integration/simulation/proxy.rs

```diff
@@ -0,0 +1,168 @@
+        }
+    }
+
+    pub fn controller(&self) -> FaultInjector {
+        FaultInjector {
+            error_before: Arc::clone(&self.error_before),
+            error_after: Arc::clone(&self.error_after),
+        }
+    }
+}
+
+struct FaultInjectorMarker;
+
+impl<H: Handler> FaultInjectorHandler<H> {
+    async fn run(&self, mut conn: Conn) -> Conn {
+        conn.insert_state(FaultInjectorMarker);
+        if *self.error_before.lock().unwrap() {
+            conn.with_status(Status::InternalServerError)
+        } else {
+            self.inner.run(conn).await
+        }
+    }
+
+    async fn before_send(&self, conn: Conn) -> Conn {
+        let mut conn = self.inner.before_send(conn).await;
+        if conn.state::<FaultInjectorMarker>().is_some() && *self.error_after.lock().unwrap() {
+            conn.set_status(Status::InternalServerError);
+            let header_names = conn
+                .response_headers()
+                .iter()
+                .map(|(name, _)| name.to_owned())
+                .collect::<Vec<_>>();
+            conn.response_headers_mut().remove_all(header_names);
+            conn.set_body("");
+        }
+        conn
+    }
+
+    fn name(&self) -> Cow<'static, str> {
+        format!("FaultInjectorHandler({})", std::any::type_name::<H>()).into()
+    }
+}
+
+/// This controls a [`FaultInjectorHandler`].
+pub(super) struct FaultInjector {
+    error_before: Arc<Mutex<bool>>,
+    error_after: Arc<Mutex<bool>>,
+}
+
+impl FaultInjector {
+    pub fn reset(&self) {
+        *self.error_before.lock().unwrap() = false;
+        *self.error_after.lock().unwrap() = false;
+    }
+
+    pub fn error_before(&self) {
+        *self.error_before.lock().unwrap() = true;
+    }
+
+    pub fn error_after(&self) {
+        *self.error_after.lock().unwrap() = true;
+    }
+}
+
+/// A [`Handler`] wrapper that inspects request and response bodies, in order to trigger test failures.
+#[derive(Handler)]
+pub(super) struct InspectHandler<H> {
+    #[handler(except = [run, before_send, name])]
+    inner: H,
+    failure: Arc<Mutex<bool>>,
+}
+
+impl<H> InspectHandler<H> {
+    pub fn new(handler: H) -> Self {
+        Self {
+            inner: handler,
+            failure: Arc::new(Mutex::new(false)),
+        }
+    }
+
+    pub fn monitor(&self) -> InspectMonitor {
+        InspectMonitor {
+            failure: Arc::clone(&self.failure),
+        }
+    }
+}
+
+struct InspectMarker;
+
+impl<H: Handler> InspectHandler<H> {
+    async fn run(&self, mut conn: Conn) -> Conn {
+        conn.insert_state(InspectMarker);
+        self.inner.run(conn).await
+    }
+
+    async fn before_send(&self, conn: Conn) -> Conn {
+        let mut conn = self.inner.before_send(conn).await;
+        if conn.state::<InspectMarker>().is_some() {
+            if conn.status() == Some(Status::Conflict) {
+                error!("409 Conflict response");
+                *self.failure.lock().unwrap() = true;
+            }
```

Comment:
Any 4xx error strikes me as fatal? Or does that break the case where report uploads are too far in the future?